src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,194 @@ internal class DiagnosticComputer
+                var computeTask = CreateComputeTask(cancellationTokenSource, cancellationToken);
+
+                // Step 3:
+                //  - Add 'cancellationTokenSource' to the list of cancellation token sources for
+                //    currently executing normal priority tasks.
+                //    This list is used in 'GetHighPriorityDiagnosticsAsync' method to cancel all the
+                //    currently executing normal priority tasks if a new high priority diagnostic
+                //    request is subsequently received.
+                lock (s_gate)
+                {
+                    Debug.Assert(!s_cancellationSourcesForNormalPriorityComputeTasks.Contains(cancellationTokenSource));
+                    s_cancellationSourcesForNormalPriorityComputeTasks.Add(cancellationTokenSource);
+                }
+
+                try
+                {
+                    // Step 4:
+                    //  - Execute the core 'computeTask' for diagnostic computation.
+                    return await computeTask.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Step 5:
+                    // Check if cancellation fired on the custom 'cancellationTokenSource' that was created for
+                    // allowing cancellation of 'computeTask' from subsequent highPriority requests.
+                    if (cancellationTokenSource.IsCancellationRequested)
+                    {
+                        // Attempt to re-execute this cancelled normal priority task
+                        // by running the loop again.
+                        continue;
+                    }
+
+                    // Propagate all other OperationCanceledExceptions up the stack.
+                    throw;
+                }
+                finally
+                {
+                    // Step 6:
+                    //  - Remove the 'cancellationTokenSource' for completed or cancelled task.
+                    //    For the case where the computeTask was cancelled, we will create a new
+                    //    'cancellationTokenSource' for the retry.
+                    lock (s_gate)
+                    {
+                        var removed = s_cancellationSourcesForNormalPriorityComputeTasks.Remove(cancellationTokenSource);
+                        Debug.Assert(removed);
+                    }
+                }
+            }
+
+            Task<SerializableDiagnosticAnalysisResults> CreateComputeTask(CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken)
+            {
+                return Task.Run(async () =>
+                {
+                    // Create a linked cancellation source to allow high priority tasks to cancel normal priority tasks.
+                    using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
```

Comment:
as per above... i don't get the need to have two CTSs... why is this not just the outer CTS's token?